### PR TITLE
feat: Support atomic replica takeover

### LINKED ISSUE #1446: Support atomic replica takeover
https://docs.google.com/document/d/1wDt_Rp4wfEX5h0cTVza1-AJOp9JrPyA5cw7CkEAjBgw/edit

### PR DESCRIPTION
**Pull request by royjacobson**
Support a safe transfer of ownership from a master to a replica.

Closes #1446

### CODE REVIEW

File: src/server/dflycmd.h [hunk 1]
```diff
@@ -159,6 +159,10 @@ class DflyCmd {
   // Switch to stable state replication.
   void StartStable(CmdArgList args, ConnectionContext* cntx);
 
+  // STARTSTABLE <syncid>
+  // Switch to stable state replication.
+  void TakeOver(CmdArgList args, ConnectionContext* cntx);
```
Comment (dranikpg) [inline src/server/dflycmd.h:164]: You forgot to change the comment

File: src/server/dflycmd.cc [hunk 2]
```diff
@@ -336,6 +339,56 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
...
+    shard->journal()->RecordEntry(0, journal::Op::PING, 0, 0, {}, true);
+
+    absl::Time start = absl::Now();
+    while (flow->last_acked_lsn < shard->journal()->GetLsn()) {
+      if (absl::ToInt64Seconds(absl::Now() - start) > timeout) {
+        VLOG(1) << "Couldn't synchronize with replica for takeover in time.";
+        status = OpStatus::TIMED_OUT;
+        return;
+      }
+      ThisFiber::SleepFor(1ms);
```
Comment (dranikpg) [inline src/server/dflycmd.cc:379]: Did you already plan any error handling? If the replica fails, we will always wait till the end, because cancellation can't run as long as you hold the replica_ptr's mutex.

Is quick error handling important here at all?

File: src/server/server_family.cc [hunk 3]
```diff
@@ -2098,9 +2133,11 @@ void ServerFamily::Register(CommandRegistry* registry) {
             << CI{"LATENCY", CO::NOSCRIPT | CO::LOADING | CO::FAST, -2, 0, 0, 0}.HFUNC(Latency)
             << CI{"MEMORY", kMemOpts, -2, 0, 0, 0}.HFUNC(Memory)
             << CI{"SAVE", CO::ADMIN | CO::GLOBAL_TRANS, -1, 0, 0, 0}.HFUNC(Save)
-            << CI{"SHUTDOWN", CO::ADMIN | CO::NOSCRIPT | CO::LOADING, -1, 0, 0, 0}.HFUNC(_Shutdown)
+            << CI{"SHUTDOWN", CO::ADMIN | CO::NOSCRIPT | CO::LOADING, -1, 0, 0, 0}.HFUNC(
+                   ShutdownCmd)
             << CI{"SLAVEOF", kReplicaOpts, 3, 0, 0, 0}.HFUNC(ReplicaOf)
             << CI{"REPLICAOF", kReplicaOpts, 3, 0, 0, 0}.HFUNC(ReplicaOf)
+            << CI{"REPLTAKEOVER", kReplicaOpts, 2, 0, 0, 0}.HFUNC(ReplTakeOver)
```
Comment (adiholden) [inline src/server/server_family.cc:2140]: Do we want to allow this command during loading?

File: src/server/server_family.cc [hunk 4]
```diff
@@ -1889,6 +1889,41 @@ void ServerFamily::ReplicaOf(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+void ServerFamily::ReplTakeOver(CmdArgList args, ConnectionContext* cntx) {
+  VLOG(1) << "Starting take over";
+  VLOG(1) << "Acquire replica lock";
+  unique_lock lk(replicaof_mu_);
+
+  float_t timeout;
```
Comment (adiholden) [inline src/server/server_family.cc:1897]: in seconds?

File: src/server/dflycmd.cc [hunk 5]
```diff
@@ -336,6 +339,56 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
...
+  TransactionGuard tg{cntx->transaction};
+  AggregateStatus status;
+
+  auto cb = [replica_ptr = replica_ptr, timeout, &status](EngineShard* shard) {
+    FlowInfo* flow = &replica_ptr->flows[shard->shard_id()];
+
+    shard->journal()->RecordEntry(0, journal::Op::PING, 0, 0, {}, true);
+
+    absl::Time start = absl::Now();
+    while (flow->last_acked_lsn < shard->journal()->GetLsn()) {
```
Comment (adiholden) [inline src/server/dflycmd.cc:373]: We should make sure that from the point we start initialize the TransactionGuard there are no more flows that write to journal. I believe that all the callbacks that are added to tx queue are ok, but what about expiring records? when expiring a record we replicate a DEL. Checkout the flow when calling RecordExpiry func

Comment (adiholden) [reply src/server/dflycmd.cc:373]: It is possible to add a callback to the shard_set (not wait for it to finish) that will delete the entry and write to the journal from ExpireIfNeeded only if entry needs to be expired.
It will also solve other correctness problems we have in replication with expiring keys.
If you want to take this task, just make sure do work on it in a different PR

File: src/server/dflycmd.cc (hunk 2 above)
Comment (royjacobson) [reply src/server/dflycmd.cc]: I added a check for cntx.IsCancelled(), which I think is enough. I don't think it should be interruptible with manual commands

File: src/server/server_family.cc (hunk 3 above)
Comment (royjacobson) [reply src/server/server_family.cc]: no, changed it.

File: src/server/db_slice.h [hunk 6]
```diff
@@ -312,6 +312,10 @@ class DbSlice {
   // Unregisted all watched key entries for connection.
   void UnregisterConnectionWatches(ConnectionState::ExecInfo* exec_info);
 
+  void SetExpireAllowed(bool is_allowed) {
+    expire_allowed_ = is_allowed;
```
Comment (adiholden) [inline src/server/db_slice.h:316]: you did not push the db_slice.cc changes

Comment (royjacobson) [reply src/server/db_slice.h:316]: I just added it in the header, no source changes.

File: src/server/transaction.cc [hunk 7]
```diff
@@ -659,7 +661,9 @@ OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
 
   // If we run only on one shard and conclude, we can avoid scheduling at all
   // and directly dispatch the task to its destination shard.
-  bool schedule_fast = (unique_shard_cnt_ == 1) && !IsGlobal() && !IsAtomicMulti();
+  bool schedule_fast = (unique_shard_cnt_ == 1) && !IsGlobal() && !IsAtomicMulti()
+      // && ServerState::tlocal()->gstate() != GlobalState::TAKEN_OVER
```
Comment (adiholden) [inline src/server/transaction.cc:665]: why did you leave this in comment?

File: src/server/server_family.cc [hunk 8]
```diff
@@ -1902,6 +1902,41 @@ void ServerFamily::ReplicaOf(CmdArgList args, ConnectionContext* cntx) {
...
+    return (*cntx)->SendError("Already a master instance");
+  auto repl_ptr = replica_;
+  CHECK(repl_ptr);
+
+  auto info = replica_->GetInfo();
+  if (!info.full_sync_done) {
+    return (*cntx)->SendError("Full sync not done");
+  }
+
+  std::error_code ec = replica_->TakeOver(ArgS(args, 0));
```
Comment (adiholden) [inline src/server/server_family.cc:1928]: TakeOver(timeout)

File: src/server/server_family.cc (hunk 4 above)
Comment (adiholden) [reply src/server/server_family.cc:1910]: add timeunit to the variable name i.e timeout_sec

File: src/server/replica.cc [hunk 9]
```diff
@@ -226,6 +226,23 @@ void Replica::Pause(bool pause) {
   sock_->proactor()->Await([&] { is_paused_ = pause; });
 }
 
+std::error_code Replica::TakeOver(std::string_view timeout) {
+  VLOG(1) << "Taking over";
+
+  std::unique_lock lk(sock_mu_);
```
Comment (adiholden) [inline src/server/replica.cc:232]: currently we use the mutex only in the connect and close of socket. Why do we need it here? worse adding a comment

File: src/server/main_service.cc [hunk 10]
```diff
@@ -695,9 +695,23 @@ bool Service::VerifyCommand(const CommandId* cid, CmdArgList args, ConnectionCon
...
+  bool allowed_by_state = true;
+  switch (etl.gstate()) {
+    case GlobalState::LOADING:
+      allowed_by_state = dfly_cntx->journal_emulated || (cid->opt_mask() & CO::LOADING);
+      break;
+    case GlobalState::SHUTTING_DOWN:
+      allowed_by_state = false;
+      break;
+    case GlobalState::TAKEN_OVER:
+      allowed_by_state = cid->name() == "REPLCONF" || cid->name() == "SAVE";
```
Comment (adiholden) [inline src/server/main_service.cc:707]: why do you return error instead of blocking the command till takeover is done?

File: src/server/journal/journal_slice.cc [hunk 11]
```diff
@@ -114,6 +114,25 @@ error_code JournalSlice::Close() {
 
 void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
   DCHECK(ring_buffer_);
+
+  if (entry.opcode != Op::NOOP) {
```
Comment (adiholden) [inline src/server/journal/journal_slice.cc:118]: why did you need to move this above?

File: src/server/db_slice.h (hunk 6 above)
Comment (adiholden) [reply src/server/db_slice.h:322]: where is expire_allowed_ used? I see you just set it

File: src/server/dflycmd.cc [hunk 12]
```diff
@@ -336,6 +353,77 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
...
+  }
+  if (timeout < 0) {
+    return (*cntx)->SendError("timeout is negative");
+  }
+
+  VLOG(1) << "Got DFLY TAKEOVER " << sync_id_str;
+
+  auto [sync_id, replica_ptr] = GetReplicaInfoOrReply(sync_id_str, rb);
+  if (!sync_id)
+    return;
```
Comment (adiholden) [inline src/server/dflycmd.cc:371]: On replica side you are calling SendNextPhaseRequest which is expecting response.

File: src/server/dflycmd.cc [hunk 13]
```diff
@@ -336,6 +353,77 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
...
+
+  unique_lock lk(replica_ptr->mu);
+  if (!CheckReplicaStateOrReply(*replica_ptr, SyncState::STABLE_SYNC, rb))
+    return;
```
Comment (adiholden) [inline src/server/dflycmd.cc:375]: also here send response

File: src/server/dflycmd.cc [hunk 14]
```diff
@@ -336,6 +353,77 @@ void DflyCmd::StartStable(CmdArgList args, ConnectionContext* cntx) {
...
+  sf_->service().SwitchState(GlobalState::ACTIVE, GlobalState::TAKEN_OVER);
+
+  absl::Duration timeout_dur = absl::Seconds(timeout);
+  absl::Time start = absl::Now();
+  AggregateStatus status;
+
+  // TODO: We should cancel blocking commands before awaiting all
+  // dispatches to finish.
+  if (!static_cast<Listener*>(listener_)->AwaitDispatches(
+          timeout_dur, [self = cntx->owner()](util::Connection* conn) { return conn != self; })) {
```
Comment (adiholden) [inline src/server/dflycmd.cc:388]: can you explain why you wait for dispatches on this specific connection?

File: tests/dragonfly/replication_test.py [hunk 15]
```diff
@@ -1044,3 +1044,63 @@ async def test_readonly_script(df_local_factory):
...
+    await wait_available_async(c_replica)
+
+    async def counter(key):
+        value = 0
+        await c_master.execute_command(f"SET {key} 0")
+        while True:
+            try:
+                value = await c_master.execute_command(f"INCR {key}")
+            except (redis.exceptions.ConnectionError, redis.exceptions.ResponseError) as e:
+                break
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1082]: I think we want to make sure in the test that we got inside this exception

File: tests/dragonfly/replication_test.py [hunk 16]
```diff
@@ -1044,3 +1044,63 @@ async def test_readonly_script(df_local_factory):
...
+        return key, value
+
+    async def block_during_takeover():
+        "Add a blocking command during takeover to make sure it doesn't block it."
+        # TODO: We need to make takeover interrupt blocking commands.
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1087]: Lets open a github issue for this and have this in seprate test mark as skip

File: tests/dragonfly/replication_test.py [hunk 17]
```diff
@@ -1044,3 +1044,63 @@ async def test_readonly_script(df_local_factory):
...
+    [2, 4],
+    [4, 2],
+    [8, 8],
+]
+
+
+@dfly_args({"proactor_threads": 2})
+@pytest.mark.parametrize("master_threads, replica_threads", cases)
+@pytest.mark.asyncio
+async def test_take_over(df_local_factory, master_threads, replica_threads):
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1060]: What other test cases did you think to add. We need more scenarios

File: src/server/journal/journal_slice.cc (hunk 11 above)
Comment (royjacobson) [reply src/server/journal/journal_slice.cc:118]: It prevents a pretty benign race - so a change can't be visible in a replica but the LSN you'd read in the master would still be the old one.

If you have multiple replicas that's more of an issue, but still conceptually I think it makes sense to put it in the journal before it's replicated

File: src/server/server_family.cc (hunk 8 above)
Comment (royjacobson) [reply src/server/server_family.cc:1928]: I actually use the unparsed string here, didn't want to parse+serialize unnecessarily

File: src/server/replica.cc (hunk 9 above)
Comment (royjacobson) [reply src/server/replica.cc:232]: good point, probably don't need it.

File: src/server/main_service.cc (hunk 10 above)
Comment (royjacobson) [reply src/server/main_service.cc:707]: Without proxy support I don't think there's much of a difference between blocking and denying, but if we want to go for it I'd rather do it in a different PR

File: src/server/dflycmd.cc (hunk 12 above)
Comment (royjacobson) [reply src/server/dflycmd.cc:371]: It's a bit confusing but `GetReplicaInfoOrReply` already sends the error on failure.

File: src/server/main_service.cc (hunk 10 above)
Comment (adiholden) [reply src/server/main_service.cc:707]: If the replica takeover failed f.e due to timeout in this implementation you are rejecting commands instead of waiting for the takeover to fail and then the commands will resume.
You are suggesting a different PR, but doesnt the transaction guard you are using solves this issue already?

File: src/server/dflycmd.cc (hunk 12 above)
Comment (adiholden) [reply src/server/dflycmd.cc:371]: Oh

File: src/server/transaction.cc (hunk 7 above)
Comment (royjacobson) [reply src/server/transaction.cc]: I needed it in a previous versions and realized it was redundant, but forgot to delete it. thanks :)

File: src/server/dflycmd.cc (hunk 14 above)
Comment (royjacobson) [reply src/server/dflycmd.cc]: added doc

File: tests/dragonfly/replication_test.py (hunk 15 above)
Comment (royjacobson) [reply tests/dragonfly/replication_test.py:1107]: done

File: src/server/db_slice.h (hunk 6 above)
Comment (royjacobson) [reply src/server/db_slice.h:328]: You're right :see_no_evil: thanks!

File: src/server/main_service.cc (hunk 10 above)
Comment (royjacobson) [reply src/server/main_service.cc:707]: It doesn't work - some commands will get executed despite the transaction guard - but I need to think about this a bit and understand why :/

Comment (adiholden) [reply src/server/main_service.cc:707]: I believe this is either commands which do not use the transaction framework which I believe we don't care about to be executed as they should not change the database data or the inline scheduling which we can be turned off

**Comment by adiholden** [comment]
@royjacobson The PR looks good, I am just waiting for your response on https://github.com/dragonflydb/dragonfly/pull/1314#discussion_r1238384181
and adding more tests I think is must, we can discuss the cases if you need help with this, though I am ok that it will be done in a different PR

File: src/server/main_service.cc (hunk 10 above)
Comment (royjacobson) [reply src/server/main_service.cc:707]: So, my problem is basically this:

1. We receive commands during the take over and schedule them into `TxQueue`.
2. They get executed when the `TransactionGuard` is released, which can happen because I do the shutdown async.

I think we could maybe flush the transaction queues and cancel the commands there, but I don't know if that's possible? Other than that, what I implemented is to just preemptively prevent commands from getting to the `TxQueue` with this.

[take_over_guard.txt](https://github.com/dragonflydb/dragonfly/files/11880250/take_over_guard.txt)

File: tests/dragonfly/replication_test.py [hunk 18]
```diff
@@ -1068,3 +1067,107 @@ async def test_readonly_script(df_local_factory):
...
+            await asyncio.sleep(1)
+            await c_replica.execute_command(f"REPLTAKEOVER 5")
+
+        _, _, *results = await asyncio.gather(delayed_takeover(), block_during_takeover(), *[counter(f"key{i}") for i in range(16)])
+        assert await c_replica.execute_command("role") == [b'master', [[b'127.0.0.1', bytes(str(replica3.port), 'ascii'), b'stable_sync']]]
+
+        for key, client_value in results:
+            replicated_value = await c_replica.get(key)
+            assert client_value == int(replicated_value)
+            replicated_value2 = await c3.get(key)
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1139]: why do we expect in the test that replica 3 will have the same values of replica 1 which received the takeover?
From what I understand in the flow we check that the lag is the same for master and the taking over replica but we dont check the lags of other replicas

File: tests/dragonfly/replication_test.py [hunk 19]
```diff
@@ -1068,3 +1067,107 @@ async def test_readonly_script(df_local_factory):
...
+                await seeder.run(target_ops=3000)
+            except (redis.exceptions.ConnectionError, ConnectionRefusedError, Exception, SystemExit):
+                pass
+
+        fill_task = asyncio.create_task(seed())
+
+        await c_replica.execute_command(f"REPLTAKEOVER 5")
+
+        assert await c_replica.execute_command("role") == [b'master', []]
+        await fill_task
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1173]: lets add a check that the master is down, that we cant connect to it

File: tests/dragonfly/replication_test.py [hunk 20]
```diff
@@ -1068,3 +1067,107 @@ async def test_readonly_script(df_local_factory):
...
+            try:
+                await seeder.run(target_ops=3000)
+            except (redis.exceptions.ConnectionError, ConnectionRefusedError, Exception, SystemExit):
+                pass
+
+        fill_task = asyncio.create_task(seed())
+
+        await c_replica.execute_command(f"REPLTAKEOVER 5")
+
+        assert await c_replica.execute_command("role") == [b'master', []]
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1172]: When the master shuts down we save a snapshot, we can in the test check that all the data in the master passed to the replica if we start the master again and compare the master data to replica data with seeder.capture() and seeder.compare

File: tests/dragonfly/replication_test.py [hunk 21]
```diff
@@ -1068,3 +1067,107 @@ async def test_readonly_script(df_local_factory):
         assert False
     except aioredis.ResponseError as roe:
         assert 'READONLY ' in str(roe)
+
+
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1071]: how about adding a test for a very small timeout that the replica will not be able to takeover?

Comment (adiholden) [reply tests/dragonfly/replication_test.py:1071]: also a test that replica failes during the takeover and master is failing during failover

Comment (adiholden) [reply tests/dragonfly/replication_test.py:1071]: lets add this test in a different PR

File: tests/dragonfly/replication_test.py (hunk 18 above)
Comment (royjacobson) [reply tests/dragonfly/replication_test.py:1142]: clarified in comment

File: tests/dragonfly/replication_test.py (hunk 20 above)
Comment (royjacobson) [reply tests/dragonfly/replication_test.py:1175]: Turns out I had a bug with shutdown snapshotting! good catch :)

File: tests/dragonfly/replication_test.py [hunk 22]
```diff
@@ -1085,25 +1085,25 @@ async def test_take_over_counters(df_local_factory, master_threads, replica_thre
...
+        replica1.client() as c1,
         master.client() as c_blocking,
         replica2.client() as c2,
         replica3.client() as c3,
     ):
-        await c_replica.execute_command(f"REPLICAOF localhost {master.port}")
+        await c1.execute_command(f"REPLICAOF localhost {master.port}")
         await c2.execute_command(f"REPLICAOF localhost {master.port}")
-        await c3.execute_command(f"REPLICAOF localhost {replica.port}")
+        await c3.execute_command(f"REPLICAOF localhost {replica1.port}")
```
Comment (adiholden) [inline tests/dragonfly/replication_test.py:1104]: you are creating replica of a replica..
This is not a feature that we claim to support, it might be working for you in the test, but it is actually no a valid configuration that we expect from user. We probably need to block this configuraiton it in our code

Comment (royjacobson) [reply tests/dragonfly/replication_test.py]: done

**Review by adiholden — APPROVED** [review]
